cdap-operational-stats-core/src/main/java/co/cask/cdap/operations/hdfs/HDFSInfo.java

```diff
@@ -57,23 +71,39 @@ public String getVersion() {
   }
 
   @Override
-  public String getWebURL() throws IOException {
-    if (HAUtil.isHAEnabled(conf, getNameService())) {
-      return getHAWebURL().toString();
-    }
-    try (DistributedFileSystem dfs = createDFS()) {
-      return rpcToHttpAddress(dfs.getUri()).toString();
+  public String getWebURL() {
+    try {
+      if (HAUtil.isHAEnabled(conf, getNameService())) {
+        URL haWebURL = getHAWebURL();
+        if (haWebURL != null) {
+          return haWebURL.toString();
+        }
+      } else {
+        try (FileSystem fs = FileSystem.get(conf)) {
+          URL webUrl = rpcToHttpAddress(fs.getUri());
+          if (webUrl != null) {
+            return webUrl.toString();
+          }
+        }
+      }
+    } catch (IOException e) {
+      LOG.warn("Error in determining HDFS URL. Web URL of HDFS will not be available in HDFS operational stats.", e);
     }
+    return null;
```

Comment:
With URI, JMX returns the following (in particular look at the weburl key). We may have to handle serialization in a specific manner if we do that.
```
{
    "hdfs": {
        "logsurl": "http://masked:50070/logs",
        "version": "2.7.1.2.3.4.7-4",
        "weburl": {
            "compositeType": {
                "className": "javax.management.openmbean.CompositeData",
                "description": "java.net.URI",
                "nameToDescription": {
                    "absolute": "absolute",
                    "authority": "authority",
                    "fragment": "fragment",
                    "host": "host",
                    "opaque": "opaque",
                    "path": "path",
                    "port": "port",
                    "query": "query",
                    "rawAuthority": "rawAuthority",
                    "rawFragment": "rawFragment",
                    "rawPath": "rawPath",
                    "rawQuery": "rawQuery",
                    "rawSchemeSpecificPart": "rawSchemeSpecificPart",
                    "rawUserInfo": "rawUserInfo",
                    "scheme": "scheme",
                    "schemeSpecificPart": "schemeSpecificPart",
                    "userInfo": "userInfo"
                },
                "nameToType": {
                    "absolute": {
                        "className": "java.lang.Boolean",
                        "description": "java.lang.Boolean",
                        "typeName": "java.lang.Boolean"
                    },
                    "authority": {
                        "className": "java.lang.String",
                        "description": "java.lang.String",
                        "typeName": "java.lang.String"
                    },
                    "fragment": {
                        "className": "java.lang.String",
                        "description": "java.lang.String",
                        "typeName": "java.lang.String"
                    },
                    "host": {
                        "className": "java.lang.String",
                        "description": "java.lang.String",
                        "typeName": "java.lang.String"
                    },
                    "opaque": {
                        "className": "java.lang.Boolean",
                        "description": "java.lang.Boolean",
                        "typeName": "java.lang.Boolean"
                    },
                    "path": {
                        "className": "java.lang.String",
                        "description": "java.lang.String",
                        "typeName": "java.lang.String"
                    },
                    "port": {
                        "className": "java.lang.Integer",
                        "description": "java.lang.Integer",
                        "typeName": "java.lang.Integer"
                    },
                    "query": {
                        "className": "java.lang.String",
                        "description": "java.lang.String",
                        "typeName": "java.lang.String"
                    },
                    "rawAuthority": {
                        "className": "java.lang.String",
                        "description": "java.lang.String",
                        "typeName": "java.lang.String"
                    },
                    "rawFragment": {
                        "className": "java.lang.String",
                        "description": "java.lang.String",
                        "typeName": "java.lang.String"
                    },
                    "rawPath": {
                        "className": "java.lang.String",
                        "description": "java.lang.String",
                        "typeName": "java.lang.String"
                    },
                    "rawQuery": {
                        "className": "java.lang.String",
                        "description": "java.lang.String",
                        "typeName": "java.lang.String"
                    },
                    "rawSchemeSpecificPart": {
                        "className": "java.lang.String",
                        "description": "java.lang.String",
                        "typeName": "java.lang.String"
                    },
                    "rawUserInfo": {
                        "className": "java.lang.String",
                        "description": "java.lang.String",
                        "typeName": "java.lang.String"
                    },
                    "scheme": {
                        "className": "java.lang.String",
                        "description": "java.lang.String",
                        "typeName": "java.lang.String"
                    },
                    "schemeSpecificPart": {
                        "className": "java.lang.String",
                        "description": "java.lang.String",
                        "typeName": "java.lang.String"
                    },
                    "userInfo": {
                        "className": "java.lang.String",
                        "description": "java.lang.String",
                        "typeName": "java.lang.String"
                    }
                },
                "typeName": "java.net.URI"
            },
            "contents": {
                "absolute": true,
                "authority": "masked:50070",
                "host": "masked",
                "opaque": false,
                "path": "",
                "port": 50070,
                "rawAuthority": "masked:50070",
                "rawPath": "",
                "rawSchemeSpecificPart": "//masked:50070",
                "scheme": "http",
                "schemeSpecificPart": "//masked:50070"
            }
        }
    }
}
```